man/ffpry.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpry 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpry \- This program generates an FFP vector of nucleic acid RY-coded features.
.SH SYNOPSIS
.BI "ffpry [" "OPTION" "] ... [" "FILE" "] ..." 
.SH DESCRIPTION
.PP
Given no options the default behavior of the program is to
generate a Feature Frequency Profile (FFP) using features of length
10 from FASTA input.  The FFP will contain the counts of nucleic acid features coded
in purine(R)/pyrmidine(Y) format.  By default the feature keys will
be printed alongside the feature counts. FASTA sequences will be read
from standard input if no options are supplied and 
.B ffpry 
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list.  In default mode features are stored in either the forward or
reverse complement direction based upon alphabetical preference.
.SH OPTIONS
.TP
.BI "\-l " LEN ", --length=" LEN	
Changes the default length of features to 
.IR "LEN" "." 
The default length is 10. Maximum length allowed in 40.
You may overide this maximum length by setting (and exporting)
the environmental variable MAX_WORD_SIZE to a different value.
.TP
.BI "\-f " FILE ", --feature-list=" FILE
.pp
Changes the behavior of the program to read a list of features from 
.IR "FILE" "."
Features can be space or newline delimited. 
As of v3.00 features are now matched,coded and reported in both the forward and reverse strand. To disable this behaviour use the 
.B -r 
option. By default the feature list is RY coded.
.TP
.BI "\-w " STR ", --mask=" STR
.pp
Use character 
.I STR
of length 
.I LEN
which specifies a character mask for the features. For example "1111101111" means to ignore the 6th character
 of a character feature. Feature maskings functions in combination with both 
.B -d 
and 
.BR "-r" "."  
.TP
.BI "\-z " K ", --rand-mask=" K
.pp	
Create a random weight mask which allows up to K mismatches.  No greater than the word length in mismatches is allowed. The weight mask is printed to standard error at the beginning of execution.  This option is cummulative with both options -d and -r.
.TP
.BI "\-s " INT ", --rand-seed=" INT
.pp
Seed the random number generator, to make randomization operations like 
.B \-z 
predictable across different runs.  This option is cummulative with both
options 
.B -d 
and 
.BR "-r" "."
The defaul seeed to the random number generator is (system time) * (process ID).
.TP
.B \-q, --quiet
.pp
Run in quiet mode.  Suppresses printing to standard error.
.TP
.B \-d, --disable
.pp	
Disable RY coding to use ATGC coding.
.TP
.B -m, --multiple
Calculate FFPs for multiple sequences in the input file.  Sequences must have their own FASTA '>' header.
.TP
.B -r, --disable-rev
Disable counting of reverse complement features.
.PP
.SH EXAMPLES
.PP
The FFP phylogeny tools are designed to be used as filters in a pipeline.
To create a key-form FFP use this syntax:
.PP
.CODE ffpry -l 3 test1.fna
.PP
This will generate an FFP of the form
.PP
.CODE RRY 4 RRR 5 YYR ...
.PP
To generate a profile of several nucleic acid sequences
.PP
.CODE ffpry -l 3 test1.fna test2.fna test3.fna
.PP
Or more succinctly
.PP
.CODE ffpry -l 3 test*.fna
.PP
The profile for each fna file will be produced on a newline delimited row.
in the following format:
.PP
.CODE AAATGA	2	ATAGTA	4	ATGGGG	1 ...
.CODE AAACGA	2	ATAGCA	1	CTGAGG	3 ...
.PP
This format is a key-value FFP.  
.PP
The utility
.B ffprwn 
which row normalizes data  must be presented with columnar data -- This
data format contains only the count values and no keys.
The columns represent a feature and each row of the FFP input corresponds
to the counts of that feature in the individual FASTA files.
This data format is recognized by several utilities, therefore you
must convert a key valued FFP into a columnar format using
.B ffpcol
.PP
For example several commands can be piped together to produce a 
divergence matrix representing FFP similarity
.PP
.CODE ffpry -l 4 test*.fna | ffpcol | ffprwn | ffpjsd
.PP
If mismatches are desired at specific positions within a feature
a mask can be applied by using the
.B \-w
option.  The format of the mask is a string of 1's and 0's which
specify whether a match is required at that postion or whether
mismatches are allowed.  For example:
.PP
.CODE ffpry -l 5 -w "10110" test*.fna 
.PP
This specifies that mismatches are allowed at the 2nd and 5th postions
in every feature.  
.PP
The 
.B \-z
option can be used to randomly create a mask with 
.B n 
number of mismatches.  The mask is printed to stderr by 
default.  The example below redirects the output of standard
error to a file.  
.PP
.CODE ffpry -l 5 -z 2 test*.fna 2> mask
.PP
To disable the default RY coding use the 
.B \-d
option which will use 4 base ATGC coding.
.PP
If the frequencies for a small set of features is desired then the
.BI "\-f " "FILE"
option can be used. 
.I FILE
specifies the name of a file containing a list of features (RY or ATGC coded).
The list of features can be tab, space or newline delimited.  
Note the length of the features must also be specified using the 
.B \-l
option as well.    
.PP
By default all of the sequence data in a particular file is merged into one
FFP. Multiple fasta records (multiple fasta sequences in a single file)  are parsed as one single sequence (for example, consider
the case of multiple contigs in a single genomes).
If your fasta files contain multiple sequences separated by '>' style records headers 
.RB "and you wish to produce FFPs for each record then specify the " "-m" " option."
.SH L\-MER FEATURES
Features are stored in a mixture of the forward and reverse complement direction. For
example, if the feature:
.PP
.CODE GTGTAGT
.PP
is encountered in the forward direction in the sequence it will be stored by
default in the hash table and reported in the output in the reverse complement direction, which is:
.PP
.CODE ACTACAC
.PP
This decision is based upon hash function precedence, whichever direction has the 
smallest hash index will be used.  This form of hashing allows for homology detection,
independent of gene strandedness.  The behavior can be disabled with the 
.B \-r
option, which will force all features to be stored in the forward direction.
Also, by default features are stored in RY coded form.  This has been shown to
improve phylogenetic signal in a number of studies especially if highly varying 
rate of mutation are suspected among different taxa.  Additionally RY coding
is much faster, but is a compromise between speed and sensitivity.  Likewise
this feature can be disabled with the 
.B \-d
option.     
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpaa(1),
.BR ffprwn(1),
.BR ffpmerge(1),
.BR ffpjsd(1),
.BR ffpcol(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpry.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/

/* Current revision: $Revision: 1.37 $ 
 * On Tag name:  $Name:  $ 
 * Date commited: $Date: 2012-02-26 01:57:45 $
 * Author: $Author: gsims $ 
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <getopt.h>
#include <math.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hashroll.h"
#include "mask.h"
#include "ffpry.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "parse_features.h"
#include "../config.h"


#define DEFAULT_WORD_LENGTH 10  /**< Default Feature length if not given by opt -l */
#define CHAR_BUFFER_SIZE 10000	/**< Default buffer size for reading sequence files */


/* Function prototypes */

static void parseFile(HASH *, FILE *);
static void loopFeatureList(HASH * h, FILE * fp);
static void loopRaw(HASH * h, FILE * fp);


/* Global variables */

char PROG_NAME[FILENAME_MAX];
char *weightVector;               /**< A mask to allow mismatches in features */
int Length = DEFAULT_WORD_LENGTH; /**< Feature length to use if not specified by opt -l */
int maxWordSize = MAX_WORD_SIZE;  /**< Maximum allowed length for opt -l */
long Buffsize = CHAR_BUFFER_SIZE; /**< File input buffer, increase value for larger genomes. opt -b can be used to change this value */


/* Option flags and option arguments */
char *cvalue = NULL;    /**< Option argument string specifying character mask, should change to w */
char *fvalue = NULL;    /**< Option argument file name specifying a file of containing features */
bool zflag = false;	/**< Create a random character mask, -z */
int zflagN = 0;	        /**< Option argument to -z */
bool sflag = false;	/**< Specify a random seed, -s */
int sflagN = 0;	        /**< Option argument to -s */
bool qflag = false;	/**< Use quiet option */
bool wflag = false;	/**< Use a character mask, -w */
bool fflag = false;	/**< Use a list of features specified in a file, -f */
bool dflag = false;	/**< Disable RY coding, -d */
bool mflag = false;	/**< Multiple sequences in one file, -m */
bool rflag = true;	/**< Do reverse complement */


char usage_str[] = "Usage: %s [OPTIONS]... [FILE]... \n\
This program generates an FFP vector of RY-coded features\n\n\
Given no options the default behavior of the program is to\n\
generate a Feature Frequency Profile (FFP) using features of length\n\
10.\n\
\t-l LEN, --length=LEN\n\
\t-f FILE, --feature-list=FILE\n\
\t-w STR, --mask=STR\n\
\t-z K, --random-mask=K\n\
\t-s INT, --rand-seed=INT\n\
\t-q, --quiet\n\
\t-d, --disable\n\
\t-r, --disable-rev\n\
\t-m, --multiple\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";



int main(int argc, char **argv)
{
    int opt;
    FILE *fp;
    HASH h;
    int option_index = 0;

    static struct option long_options[] = {
	{"length", required_argument, 0, 'l'},
	{"mask", required_argument, 0, 'w'},
	{"disable", no_argument, 0, 'd'},
	{"random-mask", required_argument, 0, 'z'},
	{"quiet", no_argument, 0, 'q'},
	{"random-seed", required_argument, 0, 's'},
	{"help", no_argument, 0, 'h'},
	{"feature-list", required_argument, 0, 'f'},
	{"multiple", no_argument, 0, 'm'},
	{"disable-rev", no_argument, 0, 'r'},
	{"version", no_argument, 0, 'v'},
	{0, 0, 0, 0}
    };


  initSignalHandlers();

  strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "l:dw:z:s:qf:hmvr",
			      long_options, &option_index)) != -1)
	switch (opt) {
	case 'l':
	    Length = atoi(optarg);
	    break;
	case 'w':
	    if ((weightVector = (char *) malloc(sizeof(char) * strlen(optarg)))==NULL) 
		fatal_at_line("%s\n",strerror(errno));
	    strcpy(weightVector, optarg);
	    wflag = !wflag;
	    break;
	case 'z':
	    zflag = !zflag;
	    zflagN = atoi(optarg);
	    break;
	case 's':
	    sflag = !sflag;
	    sflagN = atoi(optarg);
	    break;
	case 'q':
	    qflag = !qflag;
	    break;
	case 'd':
	    dflag = !dflag;
	    break;
	case 'm':
	    mflag = !mflag;
	    break;
	case 'f':
	    fflag = !fflag;
	    fvalue = optarg;
	    break;
	case 'r':
	    rflag = !rflag;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    exit(EXIT_FAILURE);
	    break;
	}

    if ( getenv("MAX_WORD_SIZE") ) 
	maxWordSize=atoi(getenv("MAX_WORD_SIZE"));	    

    if (Length > maxWordSize) 
	    fatal_msg("%d: Max Word size is: %d",Length,maxWordSize);

    if (sflag)
	srand((unsigned) sflagN);
    else
	srand((unsigned) time(NULL) * getpid());


    if (wflag) {
	if (strlen(weightVector) != Length)
	    fatal_msg("length of mask not equal to feature length.\n");
	if (!qflag)
	    warn_msg("Using feature mask: %s\n", weightVector);
    }



    if (zflag) {
	while (zflagN > Length - 1)
	    zflagN--;
	if ((weightVector = (char *) malloc(sizeof(char) * (Length + 1))) == NULL) 
		fatal_at_line("%s\n",strerror(errno));
	randweight(weightVector, Length, zflagN);
	if (!qflag)
	    warn_msg("Using feature mask: %s\n", weightVector);

    }

    //Initialize a hash
    init(&h, (zflag || wflag), nucleotide, !dflag, rflag, Length);

    //Fill with keys if restricting to a feature list
    if (fflag)
        parseFeatureList(&h,fvalue,Length,nucleotide);

    // Must now process file arguments



    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
	    	fatal_msg("%s: %s\n", *argv, strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	parseFile(&h, fp);

	if (fp != stdin)
	    fclose(fp);

    }
    while (*argv);

    //freeHash(&h);

    return EXIT_SUCCESS;
}


/**
 *
 *  Parses FASTA format FNA file
 *
 *  This function can read from stdin or from fasta files.
 *  The hash h, is populated with features differently
 *  if there is a feature list passed via the -f option.
 *
 *
 *  @param h The hash table to use.
 *  @param fp A file pointer to the nucleotide fasta file to parse.
 *  @return none
 *
 */


static void parseFile(HASH * h, FILE * fp)
{
    if (fflag)
	loopFeatureList(h, fp);
    else
	loopRaw(h, fp);

}


/**
 *
 *  Populates a hash with feature counts from a Nucleotide fasta FILE
 *  ptr restricting the search to features defined in the feature list
 *
 *  This function can read from fp which points to stdin or a fasta file.
 *  Function behaves differently if a spaced seed (or mask) is defined with
 *  -z or -w.
 *
 *
 *  @param h The hash table to use.
 *  @param fp A file pointer to the nucleotide fasta file to parse.
 *  @return none
 *
 */

static void loopFeatureList(HASH * h, FILE * fp)
{
    struct stat fattr;
    size_t optimal_size;
    ssize_t nr;
    char *buf;
    bool firstRecord=true;
    
    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fd %s: File stat error.\n",fileno(fp));

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    if ((buf = (char *) malloc(optimal_size))==NULL) 
		fatal_at_line("%s\n",strerror(errno));

    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	if (zflag || wflag)
	    chkpushatgcw(h, buf, nr,firstRecord);
	else
	    chkpushatgc(h, buf, nr,firstRecord);

	firstRecord=false;
    }
    printFeatures(h);
    free(buf);
    freeHash(h);
}


/**
 *
 *  Populates a hash with feature counts from a Nucleotide fasta FILE
 *  ptr.
 *
 *  This function can read from fp which points to stdin or a fasta file.
 *  Function behaves differently if a spaced seed (or mask) is defined with
 *  -z or -w.
 *
 *  @param h The hash table to use.
 *  @param fp A file pointer to the nucleotide fasta file to parse.
 *  @return none
 *
 */

static void loopRaw(HASH * h, FILE * fp)
{
    struct stat fattr;
    size_t optimal_size;
    ssize_t nr;
    char *buf;
    bool firstRecord=true;
    

    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fd %s : File stat error.\n",fileno(fp));

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    if ((buf = (char *) malloc(optimal_size))==NULL) 
		fatal_at_line("%s\n",strerror(errno));

    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	if (zflag || wflag)
	    pushatgcw(h, buf, nr,firstRecord);
	else
	    pushatgc(h, buf, nr,firstRecord);

	firstRecord=false;
    	
    }


    printFeatures(h);
    free(buf);
    freeHash(h);
}


